port-forwarder/fwd_tcp.go

```diff
@@ -0,0 +1,230 @@
+package port_forwarder
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"github.com/slackhq/nebula/service"
+	"golang.org/x/sync/errgroup"
+)
+
+type ForwardConfigOutgoingTcp struct {
+	ForwardConfigOutgoing
+}
+
+func (cfg ForwardConfigOutgoingTcp) ConfigDescriptor() string {
+	return fmt.Sprintf("outbound.tcp.%s.%s", cfg.localListen, cfg.remoteConnect)
+}
+
+type ForwardConfigIncomingTcp struct {
+	ForwardConfigIncoming
+}
+
+func (cfg ForwardConfigIncomingTcp) ConfigDescriptor() string {
+	return fmt.Sprintf("inbound.tcp.%d.%s", cfg.port, cfg.forwardLocalAddress)
+}
+
+type PortForwardingOutgoingTcp struct {
+	l                     *logrus.Logger
+	tunService            *service.Service
+	cfg                   ForwardConfigOutgoingTcp
+	localListenConnection *net.TCPListener
+}
+
+func (fwd PortForwardingOutgoingTcp) Close() error {
+	fwd.localListenConnection.Close()
+	return nil
+}
+
+func (cf ForwardConfigOutgoingTcp) SetupPortForwarding(
+	tunService *service.Service,
+	l *logrus.Logger,
+) (io.Closer, error) {
+	localTcpListenAddr, err := net.ResolveTCPAddr("tcp", cf.localListen)
+	if err != nil {
+		return nil, err
+	}
+	remoteTcpAddr, err := net.ResolveTCPAddr("tcp", cf.remoteConnect)
```

Comment:
I would recommend that you do this resolution when needed. For IPs it doesn't really matter, but host lookups might change after starting Nebula.

Another thing that I don't think we have to cover in this PR is looking up IPs from the Nebula network instead of using DNS

Reply:
Ah, I see. Your right. I never used it, but nebula has an experimental DNS, right?

Reply:
done. thanks